### PR TITLE
[#178] Automate release process (3)

### PR DESCRIPTION
**Pull request by klappradla**
Edition three of #178…

This is only a draft for further discussion on the topic. I did not want to override things in #184, as there are some slight adaptions of existing things there and we may want to directly compare them.

This is how a release of a random artifact from an example repo of mine looks like with the tooling in this PR:

Screenshot 2020-05-13 at 19 08 54

**tools**
- [np](https://github.com/sindresorhus/np) to create and push the git tag
- [ghr](https://github.com/tcnksm/ghr) to create a release with artifacts on Github

**additional tools**
- [wext-shipit](https://github.com/LinusU/wext-shipit) to upload releases for Chrome, Firefox and Opera
- [chrome-webstore-upload-cli](https://github.com/DrewML/chrome-webstore-upload-cli) to upload releases for Chrome
- [web-ext-sign](https://extensionworkshop.com/documentation/develop/web-ext-command-reference/#web-ext-sign) for Firefox betas
- [web-ext-submit](https://github.com/fregante/web-ext-submit) for Firefox releases

### CODE REVIEW

**Comment by klappradla** [comment]
**TIL**
- Mozilla rejects versions with a `-` in them (e.g. `v4.0.0-0`)
- `np` does not allow to input a custom version (why?)

**Comment by pmeinhardt** [comment]
Still some things to do:

1. ~Chrome version contains "unlisted" in manifest~
2. web-ext-submit still seems to fail even though the release successfully shows up on addons.mozilla.org afterwards.

```
Applying config file: ./package.json
Building web extension from /Users/pmeinhardt/Code/bitcrowd/tickety-tick/dist/firefox
Validating add-on [............................................................]
Validation results: https://addons.mozilla.org/en-US/developers/upload/dadcd36c29994780a060d14442a4b666
Signing add-on [...............................................................]
0
FAIL

o: The extension could not be signed
    at /Users/pmeinhardt/Code/bitcrowd/tickety-tick/node_modules/web-ext-submit/node_modules/web-ext/dist/web-ext.js:1:67620
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at F.execute (/Users/pmeinhardt/Code/bitcrowd/tickety-tick/node_modules/web-ext-submit/node_modules/web-ext/dist/web-ext.js:1:33970)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by klappradla** [comment]
**Update**

Turns out the `<version>unlisted` version name does not work for Chrome 😢 https://github.com/bitcrowd/tickety-tick/issues/255

File: package.json [hunk 1]
```diff
@@ -55,6 +59,7 @@
     "jsdom": "16.2.2",
     "mini-css-extract-plugin": "^0.9.0",
     "node-sass": "4.14.1",
+    "np": "https://github.com/dopecodez/np.git#54f3c8f",
```
Comment (klappradla) [inline package.json:62]: Note to self: we can probably also move back to the stable version. Turns out the version name we wanted to have isn't https://semver.org/ compatible anyways 😉

**Comment by klappradla** [comment]
Some rebase of hell 🔥

**Comment by klappradla** [comment]
I did another field experiment with this branch 👨‍🌾: I wanted to release a new version... what in the end turned out to be two versions 🤯 

Mozilla did some updates to their upload process - the manual one: one now has to manually select that one is uploading something to be released on their platform in contrast to "self hosting". Plus, they changed the name of the returned file when signing for the `unlisted` channel, so the `cp` command in our release script does not work any more 🙈

What we also did not take into account: for uploading the release to Mozilla's `listed` channel, we have to bundle it again. Otherwise it still has the `unlisted` in its name and Mozilla will reject it as there is already a release with this name.

**Comment by klappradla** [comment]
Turns out Mozilla also has a problem with our submissions for self signing the extension - which we do to download the `.xpi` file for the release on Github: they also want to look at the source code there.

I haven't yet found out how to use the `web-ext` API to also include the source code in a way that Mozilla's reviewers can find it in the respective place.

**Comment by klappradla** [comment]
Considered done (finally 🎉 ) in #296

**Pull request closed without merging**